### PR TITLE
Add AutoBackbone + ResNetBackbone

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

As #20204 is a big PR, this PR adds a part of it as a standalone PR. This PR adds the AutoBackbone class, along with one example class that it supports, namely ResNetBackbone.

## Usage

Usage is as follows:

```
from transformers import AutoImageProcessor, AutoBackbone
import torch
from PIL import Image
import requests

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

processor = AutoImageProcessor.from_pretrained("microsoft/resnet-50")
model = AutoBackbone.from_pretrained("microsoft/resnet-50", out_features=["stage1", "stage2", "stage3", "stage4"])

inputs = processor(image, return_tensors="pt")

outputs = model(**inputs)

for k,v in zip(outputs.stage_names, outputs.hidden_states):
    print(k, v.shape)
```
which prints:
```
stage1 torch.Size([1, 256, 56, 56])
stage2 torch.Size([1, 512, 28, 28])
stage3 torch.Size([1, 1024, 14, 14])
stage4 torch.Size([1, 2048, 7, 7])
```

Besides this, one can also obtain information about the channel dimension and stride for each of the requested stages, like so:

```
print(model.channels)
print(model.strides)
```

This is handy as frameworks (like MaskFormer) need to know this information at initialization time.

## To do's/questions

- [ ] We don't want `xxxBackbone` classes to be tested by all tests defined in `test_modeling_common.py`(i.e. it should probably not be part of `all_model_classes`), hence I added the class to IGNORE_NON_TESTED, and added a separate test for it. Let me know if this is ok.
- [ ] It would probably be best to not have our backbones included in the documentation from the start. For now they are just an internal part of models like DETR and MaskFormer. Could we not include them in the docs for now? Currently I'm getting:
```
Exception: The following objects are in the public init so should be documented:
 - AutoBackbone
 - ResNetBackbone
```
An alternative option could be to add backbones to the list of PRIVATE_MODELS in utils/check_repo.py for now.

### CODE REVIEW

File: src/transformers/models/resnet/modeling_resnet.py [hunk 1]
```diff
@@ -416,3 +423,84 @@ def forward(
...
+
+        # TODO calculate strides appropriately
+        current_stride = self.resnet.embedder.embedder.convolution.stride[0]
+        self.out_feature_strides = {
+            "stem": current_stride,
+            "stage1": current_stride * 2,
+            "stage2": current_stride * 2,
+            "stage3": current_stride * 2,
+            "stage4": current_stride * 2,
+        }
```
Comment (NielsRogge) [inline src/transformers/models/resnet/modeling_resnet.py:452]: cc @amyeroberts @alaradirik it would be nice to get some help here to properly compute the output strides for each of the stages.

File: src/transformers/models/resnet/modeling_resnet.py [hunk 2]
```diff
@@ -416,3 +423,84 @@ def forward(
...
         return ImageClassifierOutputWithNoAttention(loss=loss, logits=logits, hidden_states=outputs.hidden_states)
+
+
+@add_start_docstrings(
+    """
+    ResNet backbone, to be used with frameworks like DETR and MaskFormer.
+    """,
+    RESNET_START_DOCSTRING,
+)
+class ResNetBackbone(ResNetPreTrainedModel):
```
Comment (amyeroberts) [inline src/transformers/models/resnet/modeling_resnet.py:434]: Would it be the case that we implement all backbone variations of ResNet in `modeling_resnet.py`? (I'd guess so?)

Comment (NielsRogge) [reply src/transformers/models/resnet/modeling_resnet.py:434]: We would just have a single backbone per model in the ilbrary. `ResNetBackbone` should cover all cases to use ResNet as a backbone, one can provide `config.out_features` to specify which feature maps to output

Comment (amyeroberts) [reply src/transformers/models/resnet/modeling_resnet.py:434]: OK. So in the case of DETR, when we have the frozenbatchnorm replacing batchnorm, we would load the resnet backbone, and apply the transformation in `modeling_detr.py` ?

Comment (NielsRogge) [reply src/transformers/models/resnet/modeling_resnet.py:434]: Yes, that's very specific to DETR so I'd to that there.

File: src/transformers/modeling_outputs.py [hunk 3]
```diff
@@ -1263,3 +1263,19 @@ class XVectorOutput(ModelOutput):
...
+@dataclass
+class BackboneOutput(ModelOutput):
+    """
+    Base class for outputs of backbones.
+
+    Args:
+        stage_names (`tuple(str)`):
+            Names of the stages.
+        hidden_states (`tuple(torch.FloatTensor)` of shape `(batch_size, num_channels, height, width)`):
+            Hidden states (also called feature maps) of the stages.
```
Comment (sgugger) [inline src/transformers/modeling_outputs.py:1277]: Would it make more sense to call these `feature_maps` (we are creating a new API so it's the only chance we get to fix names if we want to)? I think it would be more natural to vision users no?

File: src/transformers/models/resnet/configuration_resnet.py [hunk 4]
```diff
@@ -58,6 +58,9 @@ class ResNetConfig(PretrainedConfig):
             are supported.
         downsample_in_first_stage (`bool`, *optional*, defaults to `False`):
             If `True`, the first stage will downsample the inputs using a `stride` of 2.
+        out_features (`List[str]`, *optional*):
+            If used as backbone, list of features to output. Can be any of "stem", "stage1", "stage2", "stage3",
+            "stage4",
```
Comment (sgugger) [inline src/transformers/models/resnet/configuration_resnet.py:63]:
```suggestion
            If used as backbone, list of features to output. Can be any of `"stem"`, `"stage1"`, `"stage2"`, `"stage3"`,
            `"stage4"`,
```

File: tests/models/resnet/test_modeling_resnet.py [hunk 5]
```diff
@@ -110,6 +113,20 @@ def create_and_check_for_image_classification(self, config, pixel_values, labels
...
+        model.eval()
+        result = model(pixel_values)
+
+        # verify hidden states
+        self.parent.assertEqual(len(result.hidden_states), 4)
+        self.parent.assertListEqual(result.stage_names, config.out_features)
+        self.parent.assertListEqual(list(result.hidden_states[0].shape), [3, 10, 8, 8])
+
+        # verify channels
+        self.parent.assertListEqual(model.channels, config.hidden_sizes)
```
Comment (sgugger) [inline tests/models/resnet/test_modeling_resnet.py:128]: Can we also check the strides?

File: src/transformers/models/resnet/modeling_resnet.py [hunk 6]
```diff
@@ -416,3 +423,84 @@ def forward(
...
+        outputs = self.resnet(pixel_values, output_hidden_states=True, return_dict=True)
+
+        hidden_states = outputs.hidden_states
+
+        feature_maps = []
+        for idx, stage in enumerate(self.stage_names):
+            if stage in self.out_features:
+                feature_maps.append(hidden_states[idx])
+
+        return BackboneOutput(stage_names=self.out_features, hidden_states=feature_maps)
```
Comment (sgugger) [inline src/transformers/models/resnet/modeling_resnet.py:506]: `self.out_features` is static, it doesn't change when we change the inputs, so it should just be an attribute of the backbone (as you have done), not included in the `BackboneOutput`.

**Review by sgugger — COMMENTED** [review]
Thanks for splitting the PR in smaller bits, it's far easier to review this way. My two main comments are on the backbone output:
- it shouldn't include the feature names since it's static, so having as a property of the model is enough
- it could rename `hidden_state` to `feature_maps` if we think it's a more natural name.

File: src/transformers/modeling_outputs.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/modeling_outputs.py:1277]: Makes sense, will update!

File: tests/models/resnet/test_modeling_resnet.py (hunk 5 above)
Comment (NielsRogge) [reply tests/models/resnet/test_modeling_resnet.py:127]: Actually both DETR and MaskFormer don't require the stride information, can I remove the `strides` property for now, and only keep the bare minimum?

Comment (sgugger) [reply tests/models/resnet/test_modeling_resnet.py:127]: Yes, no need to have it if it's not useful for now.

File: src/transformers/models/resnet/modeling_resnet.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/resnet/modeling_resnet.py]: I've removed the strides for now as they are not necessary for DETR and MaskFormer

**Review by amyeroberts — COMMENTED** [review]
Thanks for splitting out into this PR and for the code snippet. I really like the overall layout - it's simple and I think fits well with other library patterns.

My main question is about how the backbone is loaded and saved alongside our current models. For example, if the backbone has non-default settings - where is this information saved? Is it part of the model e.g. DETR config? 

Can we assume the backbone is always frozen, or are there cases when people might want to fine-tune it? In this case, how would the backbone weights be saved out?

**Comment by NielsRogge** [comment]
> My main question is about how the backbone is loaded and saved alongside our current models. For example, if the backbone has non-default settings - where is this information saved? Is it part of the model e.g. DETR config?

A backbone can be loaded and saved just like any other model in the library, either using a config to initialize the backbone with randomly initialized weights or using the from_pretrained method to load pre-trained weights.

> Can we assume the backbone is always frozen, or are there cases when people might want to fine-tune it? In this case, how would the backbone weights be saved out?

The backbone is always meant to be fine-tuned together with the backbone, I've not seen a case where the backbone is kept frozen for now, but but we could definitely add a `freeze` method in case people want to do that.

**Comment by NielsRogge** [comment]
@sgugger the remaining CI issue is about the fact that they are not documented, however I'd like to actually keep them away from the docs for now. However I need to keep ResNetBackbone in the main init for it to work with the Auto API.

**Comment by amyeroberts** [comment]
@NielsRogge Thanks for your answers. The backbone can definitely be saved with `save_pretrained`, my question is really about how that's bundled together with a model. For example, if I wanted a new model, with a new-non-standard backbone - is this how I would create it? 

```
from transformers import AutoConfig, AutoBackbone, AutoModelForXXX

# Modify the default backbone configuration
backbone_config = AutoConfig(backbone_repo)
backbone_config.param = new_value_0

# Modify the default model configuration
model_config = AutoConfig(model_repo)
model_config.param = new_value_1

model = AutoModelForXXX(model_config, backbone_config)
```

Or is the backbone configuration part of the model configuration?

```
from transformers import AutoConfig

model_config = AutoConfig(model_repo)
model_config["backbone"]["param"] = new_value_0
model_config.param = new_value_1

model = AutoModel(model_config)
```

or something else?

**Comment by sgugger** [comment]
@NielsRogge You can add them to [this list](https://github.com/huggingface/transformers/blob/0d0d77693f79c7f7d39bba6921cc9741f00de988/utils/check_repo.py#L665) for now.

**Comment by NielsRogge** [comment]
The way frameworks like DETR and MaskFormer can use a backbone can be as follows. In their configuration, e.g. MaskFormerConfig, they should have a `backbone_config` attribute, which can be set to ResNetConfig for instance, like so:

```
from transformers import MaskFormerConfig

backbone_config = ResNetCong(hidden_sizes=...)
config = MaskFormerConfig(backbone_config=backbone_config)
````
So yes, the backbone configuration is part of the model configuration.

Then, inside modeling_maskformer.py, one can instantiate the backbone like so:

```
from transformers import AutoBackbone

self.backbone = AutoBackbone.from_config(config.backbone_config)
```

**Comment by amyeroberts** [comment]
@NielsRogge Great - thanks for clarifying :)

**Comment by NielsRogge** [comment]
@sgugger @amyeroberts feel free to approve :)

**Review by sgugger — COMMENTED** [review]
There is one last standing comment from Amy I see.

File: src/transformers/models/resnet/modeling_resnet.py [hunk 7]
```diff
@@ -416,3 +423,70 @@ def forward(
...
+)
+class ResNetBackbone(ResNetPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.resnet = ResNetModel(config)
+
+        # note that this assumes there are always 4 stages
+        self.stage_names = ["stem", "stage1", "stage2", "stage3", "stage4"]
+        self.out_features = config.out_features
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/resnet/modeling_resnet.py:441]: Fixed this in https://github.com/NielsRogge/transformers/commit/1863e8ff5dcaf29333d5bd08f266a562d36a8ee6

**Comment by NielsRogge** [comment]
This has been fixed in https://github.com/NielsRogge/transformers/commit/1863e8ff5dcaf29333d5bd08f266a562d36a8ee6

File: src/transformers/models/resnet/configuration_resnet.py [hunk 8]
```diff
@@ -75,6 +78,8 @@ class ResNetConfig(PretrainedConfig):
     """
     model_type = "resnet"
     layer_types = ["basic", "bottleneck"]
+    # note that this assumes there are always 4 stages
+    stage_names = ["stem", "stage1", "stage2", "stage3", "stage4"]
```
Comment (amyeroberts) [inline src/transformers/models/resnet/configuration_resnet.py:82]: Shouldn't this be generated based the length of the `depths` parameter passed in the init?

**Review by sgugger — APPROVED** [review]
Good for me modulo Amy's last comment. Thanks again for your work on this!

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for implementing - exciting that we now have the `AutoBackbone`

Other than my last comment looks good to merge :)

File: src/transformers/models/resnet/configuration_resnet.py (hunk 8 above)
Comment (NielsRogge) [reply src/transformers/models/resnet/configuration_resnet.py]: Thanks for suggesting, added that!

File: src/transformers/models/resnet/modeling_resnet.py [hunk 9]
```diff
@@ -416,3 +423,69 @@ def forward(
...
+
+        >>> inputs = processor(image, return_tensors="pt")
+
+        >>> outputs = model(**inputs)
+        ```"""
+        outputs = self.resnet(pixel_values, output_hidden_states=True, return_dict=True)
+
+        hidden_states = outputs.hidden_states
+
+        feature_maps = []
```
Comment (NielsRogge) [inline src/transformers/models/resnet/modeling_resnet.py:486]: One final question; should we opt for a list or a tuple? Currently the docstring says tuple.

cc @sgugger @amyeroberts

Comment (amyeroberts) [reply src/transformers/models/resnet/modeling_resnet.py:486]: To be consistent with the other output types, I'd say tuple. Not sure how strict this is - I assume it was a design decision as the model outputs are to be considered as non-mutable?

File: src/transformers/models/resnet/modeling_resnet.py (hunk 1 above)
Comment (alaradirik) [reply src/transformers/models/resnet/modeling_resnet.py]: If we need to add something like this in the future, maybe we can use setattr() similar to the [StyleGAN2](https://github.com/NVlabs/stylegan2-ada-pytorch/blob/main/training/networks.py) implementation to retrieve the stride of each block.

File: src/transformers/models/resnet/modeling_resnet.py (hunk 9 above)
Comment (NielsRogge) [reply src/transformers/models/resnet/modeling_resnet.py:486]: Ok I'll update to tuple

**Review by alaradirik — APPROVED** [review]
Thanks for adding this! Like Amy and Sylvain said, I like the overall layout and I think it's intuitive and easy to use.

I know Sylvain already brought this up but I think outputting named feature_maps might be useful as it really makes experimentation easier. Otherwise, I agree with the other comments.

**Comment by NielsRogge** [comment]
Failing tests are unrelated (there seems to be an issue with the cache of MarkupLM and mBART50 on the CI), merging.
